### PR TITLE
Refactored module, removed obsolete files, added tests and CI.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
This PR is a major refactor of the module with replacing some procedural approaches with OOP and adding tests.

## In scope
- Preserving existing working functionality
- Adding tests
- Fixing coding standards
- Fixing deprecated code
- Support for aliases - will be provided as a separate PR.
- Replacing link paths in documents - needs to be discussed and will be provided as a separate PR.

## Changed
- Refactored core of the module to use OOP approach. Now using `Minisite` class to handle archive and asset management.
- Refactored how asset management works. 
  - Now using `MinisiteAsset` class to work with assets. This class is quite 'thin' now, but will have more functionality once alias support introduced (CRUD records in DB etc).
  - Using Drupal core `ArchiverInterface` and 2 implementation `Zip` and `Tar`. This means that there is a support for `tar` archives.
- Added separate `MinisiteValidator` class to handle specific constraints around archive contents + unit tests.
- Added tests and test scaffolding (will be useful for future tests) to make sure that:
  - Field can be created through UI
  - Field can be set as required
  - Valid archive can be uploaded
  - Invalid archive shows validation error
- Fixed numerous coding standards violations
- Fixed deprecated code
- Removed all code relevant to alias processing as it was not working and was not completely  Drupal 8 compliant. A rewrite of this functionality is coming.
- Removed all code relevant to links processing in the documents. Pending clarifications and discussions.

## How to test
There are 2 ways to test this: 1) within the project itself (same as CI does it) or 2) installing manually into custom or vanilla site.

### Option 1
1. Clone this repo
2. Checkout `8.x-refactor` branch
3. Run `./.circleci/build.sh`. If you have php and sqllite installed, a vanilla Drupal site with module enabled (running on in-built PHP server) will be provisioned.
4. Login to the site (the login link will be available in the output of the script).
5. Perform any manual testing.
Pro tip: run `./.circleci/lint.sh` to lint code and `./.circleci/test.sh` to run all available tests.

### Option 2
1. Download a tarball of `8.x-refactor` branch.
2. Install it into existing site.
3. Perform any manual testing.

### CODE REVIEW

File: .circleci/build.sh
Comment (AlexSkrypnyk) [inline .circleci/build.sh:20]: this code allows to run the same site build process locally, so it is possible to do development of this module by running `./.circleci/build.sh`

File: .circleci/config.yml
Comment (AlexSkrypnyk) [inline .circleci/config.yml:18]: lint is split into another script to allow running local linting without running tests and the other way around

File: .circleci/test.sh
Comment (AlexSkrypnyk) [inline .circleci/test.sh:11]: always use fresh DB for tests

File: composer.json
Comment (AlexSkrypnyk) [inline composer.json:10]: correct license name

File: minisite.install
Comment (AlexSkrypnyk) [inline minisite.install:35]: mainly moved code around to allow paths in Status Report. Like so:
Status_report___Drush_Site-Install

File: minisite.install
Comment (AlexSkrypnyk) [inline minisite.install:69]: we do not need this for now as the schema is required only for alias support

File: minisite.module
Comment (AlexSkrypnyk) [inline minisite.module:5]: although the file name is the same, the content is completely different.

File: minisite.module
Comment (AlexSkrypnyk) [inline minisite.module:18]: all of these, except for `minisite.theme.in`, were used (partially) for aliases, so they are removed now.

File: minisite.module
Comment (AlexSkrypnyk) [inline minisite.module:82]: upload validation callback used in widget when the archive is uploaded. Unfortunately, Drupal does not support OOP callbacks in `#upload_validation`, so this has to dangle here

File: minisite.theme.inc
Comment (AlexSkrypnyk) [inline minisite.theme.inc:8]: the theme layer stayed untouched - just code style fixes

File: minisite.theme.inc
Comment (AlexSkrypnyk) [inline minisite.theme.inc]: this was adding a theme for a minisite page, served through alias. removed for now.

File: src/Minisite.php
Comment (AlexSkrypnyk) [inline src/Minisite.php]: note that this function can be called with file being uploaded (file that has a temporary URI) and file that has already been uploaded (file leaving in the site's `files`).

File: src/Minisite.php
Comment (AlexSkrypnyk) [inline src/Minisite.php:426]: this is using our own version of ArchiverManager  - see MinisteArchiverManager for more explanation

File: src/Plugin/Field/FieldType/MinisiteItem.php
Comment (AlexSkrypnyk) [inline src/Plugin/Field/FieldType/MinisiteItem.php:169]: This was one of the 'logic leaking' places. `Minisite` is/was not an entity, so it does not know anything about saving or pre-saving.

Changed this now that a Minisite instance is instantiated and a primary entry point (index file) is retrieved.

File: src/Plugin/Field/FieldWidget/MinisiteWidget.php
Comment (AlexSkrypnyk) [inline src/Plugin/Field/FieldWidget/MinisiteWidget.php:23]: this was a copy/paste from the parent class

File: src/Plugin/Field/FieldWidget/MinisiteWidget.php
Comment (AlexSkrypnyk) [inline src/Plugin/Field/FieldWidget/MinisiteWidget.php:35]: same - not doing anything

File: src/Plugin/Field/FieldWidget/MinisiteWidget.php
Comment (AlexSkrypnyk) [inline src/Plugin/Field/FieldWidget/MinisiteWidget.php:78]: the logic here was a bit strange. 

What we want is to check that, if `file_validate_extensions` is one of our existing validators, then it should be using our own extensions to validate rather than defaults.

File: src/Plugin/Field/FieldWidget/MinisiteWidget.php
Comment (AlexSkrypnyk) [inline src/Plugin/Field/FieldWidget/MinisiteWidget.php:105]: removed as related to site alias

File: tests/src/Traits/FixtureTrait.php
Comment (AlexSkrypnyk) [inline tests/src/Traits/FixtureTrait.php:8]: this trait help to create fixture archives - we use it to test multiple permutations of the archive. 

This trait has it's own tests.

File: tests/src/Unit/MinisiteFixtureTest.php
Comment (AlexSkrypnyk) [inline tests/src/Unit/MinisiteFixtureTest.php]: tests for tests ❤️

File: .circleci/build.sh
Comment (pandaskii) [reply .circleci/build.sh:20]: Such a beautiful way

**Comment by pandaskii** [comment]
This looks fantastic, thanks for the hard work. 

Two questions, please check the attached screenshots:

1. Allowed number of values - can we force it to 1? 

Screen Shot 2019-11-22 at 11 57 10 am

2. Global blacklisted extensions and allowed file extensions

Screen Shot 2019-11-22 at 11 44 39 am

**Comment by AlexSkrypnyk** [comment]
@pandaskii thank you for reviewing. Your suggested changes are clear - will make the changes + tests.

**Review by pandaskii — APPROVED** [review]
We are good to a Drupal project release.

thanks for the great work

**Comment by AlexSkrypnyk** [comment]
@pandaskii I have another branch with major changes coming to support path aliasing. Adding those changes after release will require significantly more work. I'm actively working on finishing aliasing functionality. Could you please wait for another half-day or so and we will do a release then. Thanks

**Comment by AlexSkrypnyk** [comment]
@pandaskii The alias support has now been implemented. We are testing it internally, but it should all work now. Please test yourself and let me know when I can merge this into `8.x` branch.

Once merged, I will create a `8.x-1.0-alpha1` version of the module and will publish it on D.O.

Please note that due to the size of the module, there may still be some issues  - we can address them in `alpha2` etc.

**Pull request merged**
